src/reliability.jl

```diff
@@ -0,0 +1,72 @@
+using LinearAlgebra: Diagonal
+
+struct Reliability
+    alpha::Float64
+    dropped::Vector{Pair{Int, Float64}}
+end
+
+function Base.show(io::IO, x::Reliability)
+    println(io, "Reliability for all items: $(round(x.alpha; digits=4))")
+    println(io, "")
+    println(io, "Reliability if an item is dropped:")
+    for i ∈ x.dropped
+        println(io, "item $(i.first): $(round(i.second; digits=4))")
+    end
+end
+
+function _crombach_alpha(covmatrix::AbstractMatrix{T}) where T <: Real
+    k = size(covmatrix, 2)
+    σ = sum(covmatrix)
+    σ_ij = sum(covmatrix - Diagonal(covmatrix)) / (k * (k - 1))
+    ρ = k^2 * σ_ij / σ
+    return ρ
+end
+
+"""
+    crombach_alpha(covmatrix::AbstractMatrix{T}) where T <: Real
+
+Calculate Crombach's alpha (1951) from a covariance matrix `covmatrix` according to
+the Wikipedia formula (https://en.wikipedia.org/wiki/Cronbach%27s_alpha):
+
+```math
+\\rho = \\frac{k^2 \\bar{sigma}_{ij}}{\\sigma^2_X}
+```
+
+where k is the number of items, i.e. columns; σᵢⱼ denote the average of
+the inter-item covariances; and  σ²ₓ consists of item variances and inter-item
+covariances.
+
+Returns a `Reliability` object that holds:
+
+* `alpha`: the reliability score for all items, i.e. columns, in `comatrix`; and
+* `dropped`: A `Vector{Pair{item, score}}` reliability scores if a specific item, i.e. column, is dropped from `comatrix`.
+
+# Example
+```jldoctest
+julia> using StatsBase
+
+julia> cov_X = [10 6 6 6;
+                6 11 6 6;
+                6 6 12 6;
+                6 6 6 13];
+
+julia> crombach_alpha(cov_X)
+Reliability for all items: 0.814
+
+Reliability if an item is dropped:
+item 1: 0.75
+item 2: 0.761
+item 3: 0.771
+item 4: 0.783
+```
+"""
+function crombach_alpha(covmatrix::AbstractMatrix{T}) where T <: Real
+    alpha = _crombach_alpha(covmatrix)
+    k = size(covmatrix, 2)
+    dropped = Vector{Pair{Int64, Float64}}(undef, k)
+    @simd for i ∈ 1:k
+        reduced_covmatrix = covmatrix[1:end .!= i, 1:end .!= i]
```

Comment:
this will be very inefficient.

Seeing the code now I would do something like the following:
```
v = vec(sum(covmatrix, dims=1))
σ = sum(v)
for i in axes(v, 1)
    v[i][i] -= covmatrix[i, i]
end
σ_ij = sum(v)
```
and then you can compute crombach alpha from it. The point is that do compute dropped for variable `i` you just need to do:
```
σ - 2*v[i] - covmatrix[i,i]
σ_ij - 2*v[i]
```
as inputs to calculation. It will be much faster as you avoid almost all allocations.

Is the idea clear?

Reply:
The code is not clear. I think that we need to do allocations because we cannot change inplace the covmatrix in the for-loop. This is because we need the matrix without a row/column each time to calculate what the `crombach_alpha` would be without that row/column.

Or am I getting this wrong.

Reply:
This is what this part:
```
σ - 2*v[i] - covmatrix[i,i]
σ_ij - 2*v[i]
```
is doing. It is calculating the parts of the `crombach_alpha` parameters without `i`-th row and column

Reply:
My deepest apologies, but I could not get the idea. I changed the code in the for loop to avoid an allocation... But I could not get your ideia.

Reply:
I can see that you have changed the formula you use so I also change it in my code.

Here is the implementation - you can check that it produces the same results:
```
function ca(covmatrix::AbstractMatrix{T}) where T <: Real
    isposdef(covmatrix) || throw(ArgumentError("Covariance matrix is not positive definite!"))
    k = size(covmatrix, 2)
    k > 1  || throw(ArgumentError("Covariance matrix has only one columnn!"))
    v = vec(sum(covmatrix, dims=1))
    σ = sum(v)
    for i in axes(v, 1)
        v[i] -= covmatrix[i, i]
    end
    σ_diag = sum(i -> covmatrix[i, i], 1:k)

    alpha = k / (k - 1) * (1 - σ_diag / σ)
    if k > 2
        dropped = Vector{Pair{Int, AbstractFloat}}(undef, k)
        for i ∈ 1:k
            dropped[i] = i => (k - 1) / (k - 2) * (1 - (σ_diag - covmatrix[i,i]) / (σ - 2*v[i] - covmatrix[i,i]))
        end
    else
        # if k = 2 do not produce dropped; this has to be also
        # correctly handled in show
        dropped = Pair{Int, AbstractFloat}[]
    end
    return (alpha, dropped) # I am not using your result struct as it might change its implementation
end
```

Now consider analyzing a larger matrix, eg. with 1000 items (unlikely, but shows the difference):
```
julia> z = Matrix{Float64}(I, 1000, 1000);

julia> @time ca(z); # my code
  0.024561 seconds (1.01 k allocations: 7.668 MiB)

julia> @time crombach_alpha(z); # your code
 29.676325 seconds (13.01 k allocations: 14.888 GiB, 4.79% gc time)
```

Reply:
Oh now I get it! Thanks! you've omitted the whole `_crombach()` auxiliary function. That's great. I will add the implementation as you've suggested. Huge gains of performance...

Reply:
Ah - sorry for being brief!

I it is a bad habit of working with @nalimilan. I have been cooperating with him for over 3 years, and we seem to have converged to a state where we pass only minimal required information, and the rest is inferred from our common experience 😄.

Kind of https://www.huffingtonpost.com.au/2017/08/02/facebook-shuts-down-ai-robot-after-it-creates-its-own-language_a_23058978/.